### PR TITLE
Code refactorization to use model class instance ubiquitously

### PR DESCRIPTION
**Pull request by Dlux804**
Much of the `core/` code has been refactored to use the model instance more effectively.   `models.py` has been overhauled to simplify and use more imports to define instance methods.  Incorporation of classification models has been included but is incomplete and needs work.    **`# TODO`** are scattered throughout the code to identify places needing improvement, additions, or adjustment.  

Added export to JSON which is mostly complete but needs a fix for saving parameter grid. 

Neural networks are not yet supported but coming soon.

### CODE REVIEW

File: core/features.py [hunk 1]
```diff
@@ -66,42 +58,99 @@ def featurize(df, model_name, num_feat=None):
...
+    # print('self target array', self.target_array)
+    # print('Total counts:', self.n_tot)
+    # print('Feature input shape', self.in_shape)
+
+    self.train_features, self.test_features, self.train_target, self.test_target = train_test_split(self.feature_array, self.target_array,
+                                                                                test_size=self.test_percent,
+                                                                               random_state=self.random_seed)  # what data to split and how to do it.
+    # scale the data.  This should not hurt but can help many models
+    # TODO add this an optional feature
+    scaler = StandardScaler()
```
Comment (qle2) [inline core/features.py:122]: There are a couple of other normalization methods available in sklearn that we can test out and compare it to StandardScaler to get the best one

File: core/ingest.py [hunk 2]
```diff
@@ -18,7 +18,49 @@ def load_smiles(self, file, drop):
     # rename the column with SMILES to 'smiles'
     csv = csv.rename(columns={smiles_col.name: "smiles"})
     if drop: # drop all extra columns
-        csv = csv[['smiles', self.target]]
+        csv = csv[['smiles', self.target_name]]
 
     return csv, smiles_col
 
+def resolveID(file,column):
```
Comment (qle2) [inline core/ingest.py:25]: We should combine this function with `load_smiles`. If the script cannot find a SMILES column, resolve the ID and return the SMILES column.

File: core/junk.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
+'''
```
Comment (qle2) [inline core/junk.py:1]: This script is just here for reference correct?

File: core/models.py [hunk 4]
```diff
@@ -158,60 +162,29 @@ def store(self):
...
+    print('Now in:', os.getcwd())
+    print('Initializing model...', end=' ', flush=True)
+    # initiate model class with algorithm, dataset and target
+    model1 = MlModel('rf', 'ESOL.csv', 'water-sol', 'regression', tune=True, cv=3, opt_iter=25)
+    print('done.')
+
 # # featurize data with rdkit2d
-# model1.featurization([0])
+model1.featurize([0])
+model1.data_split(val=0.0)
```
Comment (qle2) [inline core/models.py:179]: Instead of `model1.data_split(val=0.0)` before `model1.run()`, can we modify it into `model1.run(val=0.0)`? If feel like it is disrupting the flow of the process

File: core/classifiers.py [hunk 5]
```diff
@@ -2,14 +2,20 @@
...
     """Returns model specific classifier function."""
 
     # Create Dictionary of classifiers to be called with self.algorithm as key.
 
-    classifiers = {
+    skl_cls = {
         'svc': SVC,
         'knc': KNeighborsClassifier,
-        'rfc': RandomForestClassifier
+        'rf': RandomForestClassifier
```
Comment (dickeygh) [inline core/classifiers.py:13]: In `Main.py,` the 'rfc' in line 21 (list of available classification learning algorithms) will need to be changed to 'rf'.

File: core/analysis.py [hunk 6]
```diff
@@ -269,54 +124,88 @@ def pvaM_graphs(pvaM):
...
+    # ax.legend(prop={'size': 16}, facecolor='w', edgecolor='k', shadow=True)
     fig.patch.set_facecolor('blue')  # Will change background color
     fig.patch.set_alpha(0.0)  # Makes background transparent
 
-    # plt.savefig(model_name+'-' +'.png')
-    # plt.show()
-    return plt  # Can I store a graph as an attribute to a model?
-
+    if filename is not None:
+        plt.savefig(filename + '.png')
```
[the remark this answers is not in the document]
Comment (Dlux804) [reply core/analysis.py:159]: If you come up with a good way to name it, sure.  This function is supposed to be a very general, all purpose scatter plotter.  The idea is if someone wants to keep the file, they name it, otherwise it is just `plt.show()`

File: core/analysis.py [hunk 7]
```diff
@@ -269,54 +124,88 @@ def pvaM_graphs(pvaM):
...
+    plt.show()
 
-# This function evaluates the classification model's performance on the test data using multiple different metrics.
-def classification_metrics(prediction, y_test):
 
-    accuracy = accuracy_score(prediction, y_test)
-    confusion = confusion_matrix(prediction, y_test)
-    classification = classification_report(prediction, y_test)
-    roc_auc = roc_auc_score(prediction, y_test)
+def grid_image(df, filename, molobj=True, smi='smiles'):  # list of molecules to print and substructre to align
```
[the remark this answers is not in the document]
Comment (Dlux804) [reply core/analysis.py:164]: Right now, I am not sure of how to make the grid image "pop-up" like you can a `matplotlib` graph.  So without a pop-up feature, you need to write it to file.  Notebooks are an exception to this.

File: core/classifiers.py (hunk 5 above)
Comment (Dlux804) [reply core/classifiers.py:13]: Noted.  I didn't change `main.py` yet.

File: core/features.py [hunk 8]
```diff
@@ -2,52 +2,43 @@
...
+    if feat_meth == None:  # ask for features
+        print('   {:5}    {:>15}'.format("Selection", "Featurization Method"))
+        [print('{:^15} {}'.format(*feat)) for feat in enumerate(feat_sets)];
+        feat_meth = [int(x) for x in input(
+            'Choose your features  by number from list above.  You can choose multiple with \'space\' delimiter:  ').split()]
+    selected_feat = [feat_sets[i] for i in feat_meth]
+    print("You have selected the following featurizations: ", end="   ", flush=True)
+    print(*selected_feat, sep=', ')
+    print('Calculating features...')#, end=' ')#, flush=True)
+    sleep(0.25)
```
[the remark this answers is not in the document]
Comment (Dlux804) [reply core/features.py:36]: Just to help with the command line output of the progress bar.  Sometimes the bar will show up too early, like before "calculating features...".  so I put a `sleep()` to give it time to make its print statements.

File: core/features.py (hunk 1 above)
Comment (Dlux804) [reply core/features.py:122]: Agreed.  Will add a TODO.

File: core/ingest.py (hunk 2 above)
Comment (Dlux804) [reply core/ingest.py:25]: I think it should be optional.  It may be best to optionally call this function from `load_csv` if no smiles are found or something along those lines.  This function, because it depends on the NCI API, is a bit finicky and certainly slow.

Comment (Dlux804) [reply core/ingest.py:25]: TODO added.

File: core/junk.py (hunk 3 above)
Comment (Dlux804) [reply core/junk.py:1]: Yes, it was a play thing.  Deleted.

File: core/models.py (hunk 4 above)
Comment (Dlux804) [reply core/models.py:179]: Why?  What do we gain by rolling those together?  I

File: core/name.py [hunk 9]
```diff
@@ -105,8 +105,8 @@ def name(algorithm, dataset, feat_meth, tune=False):
     else:
         tune_string = str(0)  # Else, it will be 0
     now = datetime.now()
-    date_string = now.strftime("%y%m%d%H%M%S_")  # Get date and time string
-    run_name = ''.join([date_string, algorithm_string, dataset_string, feat_string, tune_string])  # Run name
+    date_string = now.strftime("_%Y%m%d-%H%M%S")  # Get date and time string
+    run_name = ''.join([algorithm_string, dataset_string, feat_string, tune_string, date_string])  # Run name
```
[the remark this answers is not in the document]
Comment (Dlux804) [reply core/name.py:109]: Marked as TODO.

File: core/regressors.py [hunk 10]
```diff
@@ -50,9 +56,37 @@ def wrapKeras(build_func, in_shape):
...
+def get_regressor(self):
+    """Returns model specific regressor function."""
+
+    # Create Dictionary of regressors to be called with self.algorithm as key.
+    skl_regs = {
+        'ada': AdaBoostRegressor,
+        'rf': RandomForestRegressor,
+        'svr': SVR,
+        'gdb': GradientBoostingRegressor,
+        'mlp': MLPRegressor,
```
[the remark this answers is not in the document]
Comment (Dlux804) [reply core/regressors.py:68]: Leave it for now.

File: core/train.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+    print('Average R^2 = %.3f' % stats['r2_avg'], '+- %.3f' % stats['r2_std'])
+    print('Average RMSE = %.3f' % stats['rmse_avg'], '+- %.3f' % stats['rmse_std'])
+    print()
+
+def train_cls(self, n=5):
+    # set the model specific classifier function from sklearn
+
+    print("Starting model training with {} replicates.\n".format(n), end=' ', flush=True)
+    # self.regressor.fit(self.train_features, self.train_target)
```
[the remark this answers is not in the document]
Comment (Dlux804) [reply core/train.py:78]: Not really.  I have now removed it.  I tend to comment things out while I am remaking the code so I do not lose what was working originally.

**Comment by Dlux804** [comment]
@qle2 and @dickeygh Please look at it again.  Most of your comments were suggestions for future things or questions..  I added TODOs in the code for your suggestions that we will turn into Github issues after approval.

File: core/models.py (hunk 4 above)
Comment (qle2) [reply core/models.py]: At first I thought it would just be similar to calling `featurize` but that step is more like preparing the final component for ML. I see the process of splitting the data is part of `run()` process instead of a separate thing. While we don't exactly gain anything, it's just one extra thing we must remember to call for the ML process to work when it could be integrated into one command.

Comment (Dlux804) [reply core/models.py]: In the grand scheme of things, we likely wont be running these things in a line-by line fashion like at the bottom of `models.py` anyway.  We likely will run it thru `main.py` which would wrap all this up.  I few splitting the data as a place where you have to make decisions and an opportunity to analyze the effects of those decisions before running the burdensome `run()`.  You may want to see how data is distrubuted in your splits before running the model.  There are more ways to split than randomly.  If you look at MoleculeNet.ai 's summary of their datasets, you will see that they suggest how to split the data up.
